nixos/modules/services/mail/opendkim.nix

```diff
@@ -133,16 +207,51 @@ in
       after = [ "network.target" ];
       wantedBy = [ "multi-user.target" ];
 
-      preStart = ''
-        cd "${cfg.keyPath}"
-        if ! test -f ${cfg.selector}.private; then
-          ${pkgs.opendkim}/bin/opendkim-genkey -s ${cfg.selector} -d all-domains-generic-key
-          echo "Generated OpenDKIM key! Please update your DNS settings:\n"
-          echo "-------------------------------------------------------------"
-          cat ${cfg.selector}.txt
-          echo "-------------------------------------------------------------"
-        fi
-      '';
+      preStart =
+        if cfg.domainConfigs != { } then
+          ''
+            cd "${cfg.keyPath}"
+            ${lib.concatStringsSep "\n" (
+              lib.mapAttrsToList (domain: conf: ''
+                mkdir -p "${cfg.keyPath}/${domain}"
+                if ! test -f ${domain}/${conf.selector}.private; then
+                  ${pkgs.opendkim}/bin/opendkim-genkey -b ${toString cfg.keySize} -s ${conf.selector} -d ${domain} -D "${cfg.keyPath}/${domain}"
+                  echo "Generated OpenDKIM key for ${domain}! Please update your DNS settings:\n"
+                  echo "-------------------------------------------------------------"
+                  cat ${domain}/${conf.selector}.txt
+                  echo "-------------------------------------------------------------"
+                fi
+              '') cfg.domainConfigs
+            )}
+
+            cat > ${cfg.keyPath}/KeyTable <<EOF
+            ${lib.concatStringsSep "\n" (
+              lib.mapAttrsToList (
+                domain: conf:
+                "${conf.selector}._domainkey.${domain} ${domain}:${conf.selector}:${cfg.keyPath}/${domain}/${conf.selector}.private"
+              ) cfg.domainConfigs
+            )}
+            EOF
+
+            cat > ${cfg.keyPath}/SigningTable <<EOF
+            ${lib.concatStringsSep "\n" (
+              lib.mapAttrsToList (
+                domain: conf: "*@${domain} ${conf.selector}._domainkey.${domain}"
+              ) cfg.domainConfigs
+            )}
+            EOF
+          ''
+        else
+          ''
+            cd "${cfg.keyPath}"
+            if ! test -f ${cfg.selector}.private; then
+              ${pkgs.opendkim}/bin/opendkim-genkey -b ${toString cfg.keySize} -s ${cfg.selector} -d all-domains-generic-key
```

Comment:
We should probably add a package option, too.

Reply:
Hi,

If you don't mind, please let me know if you have any clarification on this subject.

Best Regards